data-prepper-core/src/main/java/org/opensearch/dataprepper/core/peerforwarder/PeerForwardingProcessorDecorator.java

```diff
@@ -69,13 +72,27 @@ public static List<Processor> decorateProcessors(
 
         final PeerForwarder peerForwarder = peerForwarderProvider.register(pipelineName, firstInnerProcessor, pluginId, identificationKeys, pipelineWorkerThreads);
 
-        return processors.stream().map(processor -> new PeerForwardingProcessorDecorator(peerForwarder, processor))
+        return processors.stream().map(processor -> new PeerForwardingProcessorDecorator(peerForwarder, processor, isPeerForwardingDisabled(processor, noPeerForwardingKeys)))
                 .collect(Collectors.toList());
     }
 
-    private PeerForwardingProcessorDecorator(final PeerForwarder peerForwarder, final Processor innerProcessor) {
+    private static boolean isPeerForwardingDisabled(Processor processor, List<String> noPeerForwardingKeys) {
+        if (processor instanceof RequiresPeerForwarding && noPeerForwardingKeys != null && noPeerForwardingKeys.size() > 0) {
+            Set<String> noPeerForwardKeys = new HashSet<>(noPeerForwardingKeys);
+            for (final String identificationKey: ((RequiresPeerForwarding) processor).getIdentificationKeys()) {
```

Comment:
With the changes I suggested above, this should become something like the following:

```
Collection<String> identificationKeys = ((RequiresPeerForwarding) processor).getIdentificationKeys());
for(Set<String> excludedIdentificationKeys : allExcludedIdentificationKeys) {
  if(excludedIdentificationKeys.equals(identificationKeys)) {
    return false;
  }
}
```